File: Sources/LedgelingsCore/Sparks.swift
import CoreGraphics
import Foundation

/// A handful of pixel stars thrown up when two creatures bump. They fly into
/// the screen, fall back toward the edge, and fade within a second.
public struct Sparks: Sendable {
    public struct Spark: Sendable {
        public var position: CGPoint
        public var velocity: CGVector
        /// Which way "down" is for this star: back toward its edge.
        public var gravity: CGVector
        public var age: Double = 0
        public var life: Double
        /// 0, 1, 2…: lets the renderer vary the colour without the logic caring.
        public var tint: Int

        public var opacity: Double { max(0, 1 - age / life) }
    }

    public var life: Double
    public var speed: ClosedRange<CGFloat>
    public var pull: CGFloat
    public private(set) var alive: [Spark] = []

    public init(life: Double = 0.7, speed: ClosedRange<CGFloat> = 70...150, pull: CGFloat = 300) {
        self.life = life
        self.speed = speed
        self.pull = pull
    }

    /// Fan `count` stars out around `inward`, within 60° either side of it.
    public mutating func burst(at point: CGPoint, inward: CGVector, count: Int = 8,
                               using rng: inout some RandomNumberGenerator) {
        let base = atan2(inward.dy, inward.dx)
        for k in 0..<count {
            let angle = base + CGFloat.random(in: -.pi / 3 ... .pi / 3, using: &rng)
            let v = CGFloat.random(in: speed, using: &rng)
            alive.append(Spark(position: point, velocity: CGVector(dx: cos(angle) * v, dy: sin(angle) * v),
                               gravity: CGVector(dx: -inward.dx * pull, dy: -inward.dy * pull),
                               life: life * Double.random(in: 0.7...1.1, using: &rng), tint: k))
        }
    }

    public mutating func update(dt: Double) {
        for i in alive.indices {
            alive[i].velocity.dx += alive[i].gravity.dx * dt
            alive[i].velocity.dy += alive[i].gravity.dy * dt
            alive[i].position.x += alive[i].velocity.dx * dt
            alive[i].position.y += alive[i].velocity.dy * dt
            alive[i].age += dt
        }
        alive.removeAll { $0.age >= $0.life }
    }
}

File: win/LedgelingsCore/Sparks.cs
namespace Ledgelings.Core;

/// <summary>A handful of pixel stars thrown up when two creatures bump. They fly into
/// the screen, fall back toward the edge, and fade within a second.</summary>
public sealed class Sparks
{
    public sealed class Spark
    {
        public Pt Position;
        public Vec Velocity;
        /// <summary>Which way "down" is for this star: back toward its edge.</summary>
        public Vec Gravity;
        public double Age;
        public double Life;
        /// <summary>0, 1, 2…: lets the renderer vary the colour without the logic caring.</summary>
        public int Tint;

        public double Opacity => Math.Max(0, 1 - Age / Life);
    }

    public double Life { get; set; }
    public (double Low, double High) Speed { get; set; }
    public double Pull { get; set; }
    private readonly List<Spark> alive = new();
    public IReadOnlyList<Spark> Alive => alive;

    public Sparks(double life = 0.7, (double Low, double High)? speed = null, double pull = 300)
    {
        Life = life;
        Speed = speed ?? (70, 150);
        Pull = pull;
    }

    /// <summary>Fan <paramref name="count"/> stars out around <paramref name="inward"/>, within 60° either side of it.</summary>
    public void Burst(Pt point, Vec inward, int count, Random rng)
    {
        var baseAngle = Math.Atan2(inward.Dy, inward.Dx);
        for (int k = 0; k < count; k++)
        {
            var angle = baseAngle + rng.Range(-Math.PI / 3, Math.PI / 3);
            var v = rng.Range(Speed);
            alive.Add(new Spark
            {
                Position = point,
                Velocity = new Vec(Math.Cos(angle) * v, Math.Sin(angle) * v),
                Gravity = new Vec(-inward.Dx * Pull, -inward.Dy * Pull),
                Life = Life * rng.Range(0.7, 1.1),
                Tint = k,
            });
        }
    }

    public void Update(double dt)
    {
        foreach (var s in alive)
        {
            s.Velocity = new Vec(s.Velocity.Dx + s.Gravity.Dx * dt, s.Velocity.Dy + s.Gravity.Dy * dt);
            s.Position = new Pt(s.Position.X + s.Velocity.Dx * dt, s.Position.Y + s.Velocity.Dy * dt);
            s.Age += dt;
        }
        alive.RemoveAll(s => s.Age >= s.Life);
    }
}
